cpp/src/arrow/util/converter.h

```diff
@@ -0,0 +1,281 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/builder.h"
+#include "arrow/chunked_array.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/checked_cast.h"
+
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+template <typename Input, typename Options>
+class ArrayConverter {
+ public:
+  using InputType = Input;
+  using OptionsType = Options;
+
+  ArrayConverter(const std::shared_ptr<DataType>& type,
+                 std::shared_ptr<ArrayBuilder> builder, Options options)
+      : sp_type_(type), sp_builder_(builder), options_(options) {}
+
+  virtual ~ArrayConverter() = default;
+  const std::shared_ptr<ArrayBuilder>& builder() const { return sp_builder_; }
+  const std::shared_ptr<DataType>& type() const { return sp_type_; }
+  Options options() const { return options_; }
+
+  virtual Status Init() { return Status::OK(); }
+  virtual Status Reserve(int64_t additional_capacity) = 0;
+  virtual Status Append(InputType value) = 0;
+  virtual Status AppendNull() = 0;
+  virtual Status Extend(Input seq, int64_t size) = 0;
```

Comment:
I don't think it makes sense to make `Extend()` a method of this interface: I wouldn't expect that all converted `InputType`s are unboxable as sequences or scalars in the way that `PyObject*` happens to be. Even in the specific case of `PyObject*` which has this property, Extend() needs to be supplemented with ExtendMasked() if a mask is present in addition to the sequence. I think Extend() should be moved to PyArrayConverter for now

Reply:
Agreed, removing.